src/ray/object_manager/ownership_based_object_directory.cc

```diff
@@ -121,16 +137,32 @@ ray::Status OwnershipBasedObjectDirectory::ReportObjectRemoved(
 void OwnershipBasedObjectDirectory::SubscriptionCallback(
     ObjectID object_id, WorkerID worker_id, Status status,
     const rpc::GetObjectLocationsOwnerReply &reply) {
+  // Objects are added to this map in SubscribeObjectLocations.
   auto it = listeners_.find(object_id);
+  // Do nothing for objects we are not listening for.
   if (it == listeners_.end()) {
     return;
   }
+  std::unordered_set<NodeID> node_ids;
+
+  if (!status.ok()) {
+    RAY_LOG(INFO) << "Worker " << worker_id << " failed to return location updates to "
+                  << "subscribers  for " << object_id << ": " << status.ToString()
+                  << ", assuming that the object was evicted.";
+    auto callbacks = it->second.callbacks;
+    for (const auto &callback_pair : callbacks) {
+      callback_pair.second(object_id, node_ids, "", NodeID::Nil(), 0);
+    }
+    return;
+  }
+
+  // Once this flag is set to true, it should never go back to false.
```

Comment:
What happened if the owner is dead? Is it just removed?

Reply:
Hmmm that's a great point. Right now, if the owner is dead, all existing subscriptions are given empty data (empty locations, 0 object size, etc.), given that we consider the object to have been evicted. But we're not storying that state on the `listener_state`, so subscriptions added in the future would be given stale state.

I think that `listener_state.subscribed` should remain `true` when the owner dies, but the `listener_state` should contain empty object locations and 0 object size, so all future subscribers will know that the object has been evicted. I'll make that change.